src/query_graph/mod.rs

```diff
@@ -393,30 +394,92 @@ impl QueryGraph {
         &self.non_trivial_followup_edges
     }
 
-    pub(crate) fn edge_for_field(
+    /// All outward edges from the given node (including self-key and self-root-type-resolution
+    /// edges). Primarily used by `@defer`, when needing to re-enter a subgraph for a deferred
+    /// section.
+    pub(crate) fn out_edges_with_federation_self_edges(
         &self,
         node: NodeIndex,
-        field: &NormalizedField,
-    ) -> Option<EdgeIndex> {
+    ) -> impl Iterator<Item = EdgeReference<QueryGraphEdge>> {
+        self.graph.edges_directed(node, Direction::Outgoing)
+    }
+
+    /// The outward edges from the given node, minus self-key and self-root-type-resolution edges,
+    /// as they're rarely useful (currently only used by `@defer`).
+    pub(crate) fn out_edges(
+        &self,
+        node: NodeIndex,
+    ) -> impl Iterator<Item = EdgeReference<QueryGraphEdge>> {
         self.graph
             .edges_directed(node, Direction::Outgoing)
-            .find_map(|edge_ref| {
-                let edge_weight = edge_ref.weight();
-                let QueryGraphEdgeTransition::FieldCollection {
-                    field_definition_position,
-                    ..
-                } = &edge_weight.transition
-                else {
-                    return None;
-                };
-                if field.data().field_position == *field_definition_position {
-                    Some(edge_ref.id())
-                } else {
-                    None
-                }
+            .filter(|edge_ref| {
+                !(edge_ref.source() == edge_ref.target()
+                    && matches!(
+                        edge_ref.weight().transition,
+                        QueryGraphEdgeTransition::KeyResolution
+                            | QueryGraphEdgeTransition::RootTypeResolution { .. }
+                    ))
             })
     }
 
+    pub(crate) fn edge_for_field(
+        &self,
+        node: NodeIndex,
+        field: &NormalizedField,
+    ) -> Option<EdgeIndex> {
+        // PORT_NOTE: The JS codebase found all matching candidates and asserted there was only one.
+        // We consider it unlikely for there to be more than one candidate given all the code paths
+        // that create edges, so we've removed the assert here and return on the first match.
```

Comment:
We could still assert (or debug assert) there isn’t a second candidate without collecting all of them:

```rust
let mut iter = /* … */;
if let Some(first) = iter.next() {
    debug_assert!(iter.next().is_none(), "unexpectedly found multiple candidates")
    // …
}

Reply:
That's a nice pattern, will add the `debug_assert!()`!